### PR TITLE
KAFKA-5345: Close KafkaClient when streams client is closed

### PR DESCRIPTION
**Pull request by rajinisivaram**

### CODE REVIEW

**Comment by rajinisivaram** [comment]
@guozhangwang Can you review, please? Thank you...

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java [hunk 1]
```diff
@@ -144,6 +144,7 @@ public void close() throws IOException {
         for (MetricsReporter metricsReporter: this.reporters) {
             metricsReporter.close();
         }
+        kafkaClient.close();
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java]: Would you not want to close this before the reporters?

**Review by ijuma — APPROVED** [review]
Thanks for the PR. Great find! Just a minor comment.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java (hunk 1 above)
Comment (rajinisivaram) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java]: @ijuma Thank you for the review. I have moved it.

**Review by ijuma — APPROVED** [review]
LGTM.

**Comment by guozhangwang** [comment]
LGTM.

**Comment by mjsax** [comment]
Why did we not add a test for this??

**Comment by ijuma** [comment]
There is hardly any logic in that method and that class is going away in the next release, right?

**Comment by mjsax** [comment]
Fair enough :) (even if I don't buy the "hardly any logic" argument -- tests are there to cover everything if possible)

**Comment by ijuma** [comment]
I agree that we should have tests for `close`. It's bad that there aren't any yet. I am happy that the class is going away given how few tests there are for it.

A separate question is how `close` methods are tested. If we add a test that verifies that every resource in every `Closeable` is actually closed by a `close` method, we'll be adding a lot of tests (we have very few tests that do that, I wrote one of them). Cost/benefit ratio is a bit unclear, but it may be worth having a discussion about (I think the biggest value would be to ensure that people think about what needs to be closed when writing the code in the first place).

**Comment by ijuma** [comment]
A better solution would probably be to have some automated leak detection tests.

**Comment by jvwilge** [comment]
I can confirm this patch is working, thanks for the quick fix!

Is it possible to also apply this patch to 0.10.2.2?

Maybe my example can be used to make an (integration) test for fix. Without the fix there are 850 open file handles after 5 minutes and with the fix it's only 70.

**Comment by ijuma** [comment]
Yes, it's fine to backport this. Synced with @rajinisivaram offline and he will push the change to 0.10.2 as well. She will also update the target version in the JIRA so that this is clear.

**Pull request closed without merging**